cli/js/format_error_test.ts

```diff
@@ -0,0 +1,236 @@
+    lineNumber: 1000,
+    scriptResourceName: "foo.ts",
+    startColumn: 1,
+    endColumn: 5,
+    code: 4000
+  }
+];
+
+const expected04 = `[1;31merror[0m[1m TS4000[0m: Example error
+
+► [38;5;14mfoo.ts[0m:[38;5;11m1001[0m:[38;5;11m2[0m
+
+[47;30m1001[0m abcdefghijklmnopqrstuv
+[47;30m    [0m [31m ~~~~[0m
+
+  Related information
+
+    ► [38;5;14mfoo.ts[0m:[38;5;11m100[0m:[38;5;11m4[0m
+
+    [47;30m100[0m 1234567890
+    [47;30m   [0m [38;5;14m   ~~~~[0m
+
+`;
+
+const fixture05: Deno.DiagnosticItem[] = [
+  {
+    message: "Error 001",
+    category: Deno.DiagnosticCategory.Error,
+    sourceLine: "abcdefghijklmnopqrstuv",
+    lineNumber: 1000,
+    scriptResourceName: "foo.ts",
+    startColumn: 1,
+    endColumn: 5,
+    code: 4000
+  },
+  {
+    message: "Error 002",
+    category: Deno.DiagnosticCategory.Error,
+    sourceLine: "abcdefghijklmnopqrstuv",
+    lineNumber: 1000,
+    scriptResourceName: "foo.ts",
+    startColumn: 1,
+    endColumn: 5,
+    code: 4001
+  },
+  {
+    message: "Error 003",
+    category: Deno.DiagnosticCategory.Error,
+    sourceLine: "abcdefghijklmnopqrstuv",
+    lineNumber: 1000,
+    scriptResourceName: "foo.ts",
+    startColumn: 1,
+    endColumn: 5,
+    code: 4002
+  },
+  {
+    message: "Error 004",
+    category: Deno.DiagnosticCategory.Error,
+    sourceLine: "abcdefghijklmnopqrstuv",
+    lineNumber: 1000,
+    scriptResourceName: "foo.ts",
+    startColumn: 1,
+    endColumn: 5,
+    code: 4003
+  }
+];
+
+const expected05 = `[1;31merror[0m[1m TS4000[0m: Error 001
+
+► [38;5;14mfoo.ts[0m:[38;5;11m1001[0m:[38;5;11m2[0m
+
+[47;30m1001[0m abcdefghijklmnopqrstuv
+[47;30m    [0m [31m ~~~~[0m
+
+[1;31merror[0m[1m TS4001[0m: Error 002
+
+► [38;5;14mfoo.ts[0m:[38;5;11m1001[0m:[38;5;11m2[0m
+
+[47;30m1001[0m abcdefghijklmnopqrstuv
+[47;30m    [0m [31m ~~~~[0m
+
+[1;31merror[0m[1m TS4002[0m: Error 003
+
+► [38;5;14mfoo.ts[0m:[38;5;11m1001[0m:[38;5;11m2[0m
+
+[47;30m1001[0m abcdefghijklmnopqrstuv
+[47;30m    [0m [31m ~~~~[0m
+
+[1;31merror[0m[1m TS4003[0m: Error 004
+
+► [38;5;14mfoo.ts[0m:[38;5;11m1001[0m:[38;5;11m2[0m
+
+[47;30m1001[0m abcdefghijklmnopqrstuv
+[47;30m    [0m [31m ~~~~[0m
+
+
+Found 4 errors.
+`;
+
+test(function formatDiagnosticBasic() {
+  const actual = formatDiagnostics(fixture01);
+  assertEquals(actual, expected01);
+});
+
+test(function formatDiagnosticColSpan() {
+  const actual = formatDiagnostics(fixture02);
+  assertEquals(actual, expected02);
+});
+
+test(function formatDiagnosticMessageChain() {
+  const actual = formatDiagnostics(fixture03);
+  assertEquals(actual, expected03);
+});
+
+test(function formatDiagnosticRelatedInfo() {
+  const actual = formatDiagnostics(fixture04);
+  assertEquals(actual, expected04);
+});
+
+test(function formatDiagnosticRelatedInfo() {
+  const actual = formatDiagnostics(fixture05);
+  assertEquals(actual, expected05);
+});
```

Comment:
No test for malformed diagnostic? 

```ts
formatDiagnostics({ hello: 123 });
```

Reply:
It panics at the moment, because the `DiagnosticItem.from_json_value()` liberally uses `.unwrap()` when picking values out of the JSON structure.  It would require a lot of refactoring to bubble that back up to the OP, so the OP could return something that isn't panicky (unless there is an easier way I am not aware of).

I would prefer to open another issue to deal with refactoring that and returning something non-panicky.

Reply:
I've added a test. We need to eventually make Diagnostic parsing use serde_json - added a TODO for that.